lib/agent0/agent0/utilities/predict.py

```diff
@@ -0,0 +1,234 @@
+    TradeDeltas
+        The predicted deltas of base, bonds, and shares.
+
+    """
+    if pool_state is None:
+        pool_state = deepcopy(hyperdrive_interface.current_pool_state)
+    spot_price = hyperdrive_interface.calc_spot_price(pool_state)
+    price_discount = FixedPoint(1) - spot_price
+    curve_fee = pool_state.pool_config.fees.curve
+    governance_fee = pool_state.pool_config.fees.governance_lp
+    share_price = hyperdrive_interface.current_pool_state.pool_info.vault_share_price
+    if base is not None and bonds is None:
+        if for_pool is False:
+            base_needed = base
+        else:
+            # scale up input to account for fees
+            base_needed = base / (FixedPoint(1) - price_discount * governance_fee)
+    elif bonds is not None and base is None:
+        # we need to calculate base_needed
+        bonds_needed = bonds
+        shares_needed = hyperdrive_interface.calc_shares_in_given_bonds_out_up(bonds_needed)
+        # scale down output to account for fees
+        if for_pool is False:
+            shares_needed /= FixedPoint(1) - price_discount * curve_fee
+        else:
+            shares_needed /= FixedPoint(1) - price_discount * curve_fee * governance_fee
+        share_price_on_next_block = share_price * (
+            FixedPoint(1) + hyperdrive_interface.get_variable_rate(pool_state.block_number) / FixedPoint(BLOCKS_IN_YEAR)
+        )
+        base_needed = shares_needed * share_price_on_next_block
+    else:
+        raise ValueError("predict_long(): Need to specify either bonds or base, but not both.")
+    # continue with common logic, now that we have base_needed
+    assert base_needed is not None
+    bonds_after_fees = hyperdrive_interface.calc_open_long(base_needed)
+    if verbose:
+        logging.info("predict_long(): bonds_after_fees is %s", bonds_after_fees)
+    bond_fees = bonds_after_fees * price_discount * curve_fee
+    bond_fees_to_pool = bond_fees * (FixedPoint(1) - governance_fee)
+    bond_fees_to_gov = bond_fees * governance_fee
+    bonds_before_fees = bonds_after_fees + bond_fees_to_pool + bond_fees_to_gov
+    if verbose:
+        logging.info("predict_long(): bonds_before_fees is %s", bonds_before_fees)
+        logging.info("predict_long(): bond_fees_to_pool is %s", bond_fees_to_pool)
+        logging.info("predict_long(): bond_fees_to_gov is %s", bond_fees_to_gov)
+    predicted_delta_bonds = -bonds_after_fees - bond_fees_to_gov
+    # gov_scaling factor is the ratio by which we lower the change in base and increase the change in shares
+    # this is done to take into account the effect of the governance fee on pool reserves
+    gov_scaling_factor = FixedPoint(1) - price_discount * curve_fee * governance_fee
+    predicted_delta_base = base_needed * gov_scaling_factor
+    predicted_delta_shares = base_needed / share_price * gov_scaling_factor
+    if verbose:
+        logging.info("predict_long(): predicted pool delta bonds is %s", predicted_delta_bonds)
```

Comment:
Can we get rid of these verbose blocks? I think an alternative, if you think an end-user would want to know, is to use `logging.debug` but even then keep in mind our recommended workflow is to use the debugger & breakpoints to inspect variables, so this still seems like it's unnecessary.